### PR TITLE
Prop

### PR DESCRIPTION
**Pull request by cr3a7ure**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1079
| License       | MIT
| Doc PR        | 

I have added the ApiResource 'type' and ApiProperty 'vocabType'.
As I've mentioned in the #1079 , when parsing a json-ld graph, using the exact vocabulary terms as '@id' for classes and properties, creates some issues.
In my test scenario I am parsing my graphs and create a new one, combining many of those classes.
'type' and 'vocabType' are used to describe that this class is of type schema:Class, with an IRI - '@id' of 'localVocab:Class'
Currently those keywords are not used. I'll try to submit the new Hydra documentation normalizer without breaking anything.
This is my first PR, I tried to follow the rules.

@dunglas

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@
 /vendor/
 /tests/Fixtures/app/cache/*
 /tests/Fixtures/app/logs/*
+composer.phar
```
Comment (sroze) [inline .gitignore]: Shouldn't be there but in your global gitignore 😉

File: src/Metadata/Property/PropertyMetadata.php [hunk 2]
```diff
@@ -30,12 +30,13 @@
...
     private $required;
     private $iri;
+    private $vocabType;
     private $identifier;
     private $childInherited;
     private $attributes;
     private $subresource;
 
-    public function __construct(Type $type = null, string $description = null, bool $readable = null, bool $writable = null, bool $readableLink = null, bool $writableLink = null, bool $required = null, bool $identifier = null, string $iri = null, $childInherited = null, array $attributes = null, SubresourceMetadata $subresource = null)
+    public function __construct(Type $type = null, string $description = null, bool $readable = null, bool $writable = null, bool $readableLink = null, bool $writableLink = null, bool $required = null, bool $identifier = null, string $iri = null, string $vocabType = null, $childInherited = null, array $attributes = null, SubresourceMetadata $subresource = null)
```
Comment (sroze) [inline src/Metadata/Property/PropertyMetadata.php]: Changing the argument order is a backward compatible break that we can't add without a new major version release. One of way adding arguments without BC-break is to add it at the end of the constructor, with a default `null` value.

File: src/Metadata/Resource/ResourceMetadata.php [hunk 3]
```diff
@@ -23,15 +23,17 @@
     private $shortName;
     private $description;
     private $iri;
+    private $type;
     private $itemOperations;
     private $collectionOperations;
     private $attributes;
 
-    public function __construct(string $shortName = null, string $description = null, string $iri = null, array $itemOperations = null, array $collectionOperations = null, array $attributes = null)
+    public function __construct(string $shortName = null, string $description = null, string $iri = null, string $type = null, array $itemOperations = null, array $collectionOperations = null, array $attributes = null)
```
Comment (sroze) [inline src/Metadata/Resource/ResourceMetadata.php]: Same comment that for `PropertyMetadata`'s constructor.

File: src/Metadata/Resource/ResourceMetadata.php [hunk 4]
```diff
@@ -113,6 +115,31 @@ public function withIri(string $iri): self
...
+    }
+
+    /**
+     * Returns a new instance with the given type.
+     *
+     * @param string $type
+     *
+     * @return self
+     */
+    public function withType(string $type): self
```
Comment (sroze) [inline src/Metadata/Resource/ResourceMetadata.php]: Can't this be `null` as well?

**Comment by sroze** [comment]
Thanks for your PR @cr3a7ure. What do you mean by this?
> I'll try to submit the new Hydra documentation normalizer without breaking anything.

Also, could you try to find a better name for your PR and maybe squashing your commits with a name that would helps us to understand what you've tried to do?

**Comment by Simperfit** [comment]
@cr3a7ure Are you going forward with this ?

**Comment by cr3a7ure** [comment]
@Simperfit I have stopped all development since November.
My test case seems live at http://149.202.40.241/schydra/playground.html

**Comment by soyuka** [comment]
closing as old, lmk if I need to reopen

**Pull request closed without merging**